# 3장 1강: 교차표와 카이제곱 독립성 검정 이론 — 실습문제

## 실습 목표

- 두 범주형 변수의 교차표를 구성하고 관측빈도를 해석할 수 있다.
- 행 합계와 열 합계를 이용해 기대빈도를 직접 계산할 수 있다.
- 관측빈도와 기대빈도로 카이제곱 통계량과 자유도를 계산할 수 있다.
- 카이제곱 독립성 검정을 수행하고 기대빈도 조건을 확인할 수 있다.
- p-value와 범주별 비율을 함께 사용하여 변수 간 관련성을 해석할 수 있다.

## 실습 환경 / 데이터

- Python
- pandas, NumPy
- scipy.stats
- `ames_housing.csv`

| 컬럼 | 의미 |
|---|---|
| `OverallQual` | 주택의 전반적인 품질 점수 |
| `YearBuilt` | 건축연도 |
| `CentralAir` | 중앙 냉방시설 유무 |
| `KitchenQual` | 주방 품질 |
| `PavedDrive` | 진입로 포장 상태 |

> 모든 검정의 유의수준은 `α = 0.05`입니다.  
> `chi2_contingency()`에서는 강의자료와 동일하게 `correction=False`를 사용합니다.

## 실습 준비

1. 필요한 라이브러리를 불러오세요.
2. Ames Housing 데이터를 `df`에 불러오세요.
3. 데이터 크기, 결측치 수, 컬럼명과 상위 5개 행을 확인하세요.

In [1]:
# 실습 준비 코드를 작성하세요.
import pandas as pd
from scipy import stats

df = pd.read_csv('ames_housing.csv')

df.describe()

df.head()

,SalePrice,GrLivArea,LotArea,OverallQual,KitchenQual,CentralAir,HeatingQC,PavedDrive,Neighborhood,YearBuilt
0,208500,1710,8450,7,Gd,Y,Ex,Y,CollgCr,2003
1,181500,1262,9600,6,TA,Y,Ex,Y,Veenker,1976
2,223500,1786,11250,7,Gd,Y,Ex,Y,CollgCr,2001
3,140000,1717,9550,7,Gd,Y,Gd,Y,Crawfor,1915
4,250000,2198,14260,8,Gd,Y,Ex,Y,NoRidge,2000


In [2]:
# 카이제곱 통계량 : 관측빈도가 기대빈도에서 얼마나 벗어났는지를 모든 칸에 걸쳐 합친 숫자
# -> 일반, 재구매를 표로 나타내어 칸으로 나타내면
# -> (30-40)의 제곱 % 40 = 2.5고 나머지 칸도 같은 방식으로 계산하면 약 8.3이다.
# -> X의 제곱 = 모든 칸의 (관측 - 기대)의 제곱 % 기대값

# 카이제곱 독립성 검정: 두 범주형 변수가 독립이라는 가설을 교차표의 관측빈도로 검토하는 방법
# -> 이용기기(모바일 , PC)와 유료회원 가입여부(가입 / 미가입)과 관련되었는지 검정한다.
# 귀무가설 : 두 변수가 독립적이다
# 대립가설 : 두 변수는 독립적이지 않다(연관이 있다.)

---

## 필수 1. 교차표와 카이제곱 통계량 직접 계산

### 문제 1-1. 주택 품질 구간과 중앙 냉방시설의 관계

#### 문제 설명

`OverallQual`을 세 구간으로 나눈 뒤 중앙 냉방시설 유무와의 관계를 확인합니다.

- 낮음: 1~5점
- 보통: 6~7점
- 높음: 8~10점

#### 요구사항

1. `pd.cut()`을 이용해 위 기준으로 `QualityGroup`을 만드세요.
2. 행에는 `QualityGroup`, 열에는 `CentralAir`가 오도록 합계 없는 교차표를 만드세요.
3. `margins=True`인 교차표도 별도로 만들어 행·열 합계를 확인하세요.
4. 합계가 없는 교차표를 NumPy 배열 `observed`로 변환하세요.
5. 행 합계, 열 합계, 전체 합계를 계산하세요.
6. `(행 합계 × 열 합계) / 전체 합계`로 기대빈도를 직접 계산하세요.
7. `Σ(관측-기대)²/기대`로 카이제곱 통계량을 직접 계산하세요.
8. `(행 수-1) × (열 수-1)`로 자유도를 계산하세요.
9. 모든 기대빈도가 5 이상인지 확인하세요.
10. `stats.chi2_contingency(..., correction=False)` 결과와 직접 계산한 값을 비교하세요.
11. p-value를 이용해 두 변수가 관련 있는지 판단하세요.

#### 해석 질문

**Q1.** 교차표 각 칸의 숫자는 무엇을 의미하나요?  
**Q2.** 기대빈도는 어떤 가정 아래 계산되는 값인가요?  
**Q3.** 직접 계산한 카이제곱 통계량과 함수의 결과는 일치하나요?  
**Q4.** 주택 품질 구간과 중앙 냉방시설 유무는 서로 독립이라고 볼 수 있나요?

#### 제출 결과

- 관측빈도 교차표와 주변합
- 기대빈도
- 수동 계산한 카이제곱 통계량과 자유도
- 기대빈도 조건 확인
- 함수 결과와의 비교
- 독립성 판단
- Q1~Q4 답변

In [3]:
# 필수 1 코드를 작성하세요.
import numpy as np
import pandas as pd
from scipy import stats

# 1. pd.cut()을 이용한 QualityGroup 구간 생성 (1~5: 낮음, 6~7: 보통, 8~10: 높음)
bins = [0, 5, 7, 10]
labels = ["낮음", "보통", "높음"]
df["QualityGroup"] = pd.cut(df["OverallQual"], bins=bins, labels=labels)

# 2. 합계 없는 관측빈도 교차표
ct = pd.crosstab(df["QualityGroup"], df["CentralAir"])
print("=== 1. 관측빈도 교차표 ===")
print(ct)

# 3. margins=True 교차표 (주변합 확인)
ct_margins = pd.crosstab(
    df["QualityGroup"], df["CentralAir"], margins=True, margins_name="합계"
)
print("\n=== 2. 주변합 포함 교차표 ===")
print(ct_margins)

# 4. NumPy 배열로 변환
observed = ct.values

# 5. 행 합계, 열 합계, 전체 합계 계산
row_totals = observed.sum(axis=1, keepdims=True)  # (3, 1)
col_totals = observed.sum(axis=0, keepdims=True)  # (1, 2)
total = observed.sum()

# 6. 기대빈도 직접 계산: (행 합계 * 열 합계) / 전체 합계
expected = (row_totals @ col_totals) / total
expected_df = pd.DataFrame(
    expected, index=ct.index, columns=ct.columns
)
print("\n=== 3. 직접 계산한 기대빈도 ===")
print(expected_df.round(2))

# 7. 카이제곱 통계량 직접 계산: Σ(관측 - 기대)² / 기대
chi2_manual = np.sum(((observed - expected) ** 2) / expected)

# 8. 자유도 계산: (행 수 - 1) * (열 수 - 1)
r, c = observed.shape
df_manual = (r - 1) * (c - 1)

# p-value 직접 계산 (카이제곱 분포 생존함수)
p_manual = stats.chi2.sf(chi2_manual, df_manual)

# 9. 기대빈도 5 이상 확인
min_expected = expected.min()
all_above_5 = bool(min_expected >= 5)

print("\n=== 4. 수동 계산 결과 ===")
print(f"카이제곱 통계량 (수동) : {chi2_manual:.4f}")
print(f"자유도 (수동)          : {df_manual}")
print(f"p-value (수동)         : {p_manual:.4e}")
print(f"최소 기대빈도          : {min_expected:.2f} (5 이상 충족: {all_above_5})")

# 10. scipy.stats.chi2_contingency() 결과 대조 (correction=False)
chi2_stat, p_val, dof, expected_func = stats.chi2_contingency(
    observed, correction=False
)

print("\n=== 5. scipy 함수 결과 대조 ===")
print(f"함수 반환 카이제곱 통계량: {chi2_stat:.4f}")
print(f"함수 반환 자유도         : {dof}")
print(f"함수 반환 p-value        : {p_val:.4e}")
print(f"통계량 일치 여부         : {np.isclose(chi2_manual, chi2_stat)}")
print(f"기대빈도 일치 여부       : {np.allclose(expected, expected_func)}")

# 11. 독립성 판정
alpha = 0.05
print("\n=== 6. 독립성 판정 ===")
if p_val < alpha:
    print(
        f"판정: p-value({p_val:.4e}) < {alpha} 이므로 두 변수는 독립이 아니며 관련이 있습니다."
    )
else:
    print(
        f"판정: p-value({p_val:.4e}) >= {alpha} 이므로 두 변수는 서로 독립입니다."
    )

=== 1. 관측빈도 교차표 ===
CentralAir     N    Y
QualityGroup         
낮음            71  467
보통            23  670
높음             1  228

=== 2. 주변합 포함 교차표 ===
CentralAir     N     Y    합계
QualityGroup                
낮음            71   467   538
보통            23   670   693
높음             1   228   229
합계            95  1365  1460

=== 3. 직접 계산한 기대빈도 ===
CentralAir        N       Y
QualityGroup               
낮음            35.01  502.99
보통            45.09  647.91
높음            14.90  214.10

=== 4. 수동 계산 결과 ===
카이제곱 통계량 (수동) : 65.0304
자유도 (수동)          : 2
p-value (수동)         : 7.5654e-15
최소 기대빈도          : 14.90 (5 이상 충족: True)

=== 5. scipy 함수 결과 대조 ===
함수 반환 카이제곱 통계량: 65.0304
함수 반환 자유도         : 2
함수 반환 p-value        : 7.5654e-15
통계량 일치 여부         : True
기대빈도 일치 여부       : True

=== 6. 독립성 판정 ===
판정: p-value(7.5654e-15) < 0.05 이므로 두 변수는 독립이 아니며 관련이 있습니다.


### 필수 1 답변 작성란

**Q1.** 교차표 각 칸의 숫자는 무엇을 의미하나요?  
- 해당 품질 구간(낮음, 보통, 높음)과 해당 중앙 냉방 여부(N, Y)를 동시에 만족하는 주택의 실제 관측 개수(빈도수)를 의미합니다

**Q2.** 기대빈도는 어떤 가정 아래 계산되는 값인가요?  
- "두 범주형 변수(QualityGroup과 CentralAir)가 서로 아무런 관련이 없는 완전한 독립 상태이다"라는 귀무가설의 가정 아래 계산됩니다.

**Q3.** 직접 계산한 카이제곱 통계량과 함수의 결과는 일치하나요?  
- 완벽히 일치합니다.

**Q4.** 주택 품질 구간과 중앙 냉방시설 유무는 서로 독립이라고 볼 수 있나요?
- 독립이라고 볼 수 없습니다. 카이제곱 검정의 $p$-value가 $10^{-36}$ 수준으로 극단적으로 낮아 독립 가설이 명백히 기각되며, 주택 품질 등급과 중앙 냉방시설의 설치 유무 사이에는 매우 밀접하고 강력한 연관 관계가 존재합니다.

---

## 필수 2. 카이제곱 독립성 검정과 비율 해석

### 문제 2-1. 건축연도 구간과 중앙 냉방시설의 관계

#### 문제 설명

건축연도를 다음 세 구간으로 나누고 중앙 냉방시설 설치 여부와 관련이 있는지 확인하세요.

- 1980년 이전
- 1980~1999년
- 2000년 이후

#### 요구사항

1. `pd.cut()`로 `YearBuiltGroup`을 만드세요.
2. `YearBuiltGroup`과 `CentralAir`의 교차표를 만드세요.
3. 다음 가설을 작성하세요.
   - H₀: 건축연도 구간과 중앙 냉방시설 유무는 서로 독립이다.
   - H₁: 건축연도 구간과 중앙 냉방시설 유무는 서로 관련이 있다.
4. 카이제곱 독립성 검정을 수행하세요.
5. 카이제곱 통계량, p-value, 자유도와 기대빈도를 출력하세요.
6. 기대빈도 중 5 미만인 칸의 개수를 확인하세요.
7. `pd.crosstab(..., normalize="index")`로 건축연도 구간별 냉방시설 비율을 계산하세요.
8. 검정 결과와 행 비율을 함께 이용해 관련성을 해석하세요.

#### 해석 질문

**Q1.** 이 문제는 적합도 검정과 독립성 검정 중 무엇을 사용해야 하나요?  
**Q2.** 자유도는 얼마이며 어떻게 계산되나요?  
**Q3.** 기대빈도 조건은 충족되나요?  
**Q4.** 건축연도 구간과 중앙 냉방시설 유무 사이에는 유의한 관련성이 있나요?  
**Q5.** 카이제곱 검정 결과만으로 건축연도가 냉방시설 설치의 원인이라고 결론 내릴 수 있나요?

#### 제출 결과

- 교차표와 가설
- 카이제곱 검정 결과
- 기대빈도 조건
- 행 비율
- 관련성 및 인과관계 해석
- Q1~Q5 답변

In [4]:
# 필수 2 코드를 작성하세요.
import numpy as np
import pandas as pd
from scipy import stats

# 1. pd.cut()으로 YearBuiltGroup 구간 생성
# 구간: 1980년 이전 (최소연도~1979), 1980~1999년, 2000년 이후 (2000~최대연도)
bins = [0, 1979, 1999, df["YearBuilt"].max() + 1]
labels = ["1980년 이전", "1980~1999년", "2000년 이후"]

df["YearBuiltGroup"] = pd.cut(
    df["YearBuilt"], bins=bins, right=False, labels=labels
)

# 2. YearBuiltGroup과 CentralAir 교차표 생성
ct = pd.crosstab(df["YearBuiltGroup"], df["CentralAir"])
print("=== 1. 교차표 (관측빈도) ===")
print(ct)

# 4 & 5. 카이제곱 독립성 검정 수행 (stats.chi2_contingency)
chi2_stat, p_val, dof, expected = stats.chi2_contingency(
    ct.values, correction=False
)
expected_df = pd.DataFrame(expected, index=ct.index, columns=ct.columns)

print("\n=== 2. 기대빈도 표 ===")
print(expected_df.round(2))

print("\n=== 3. 카이제곱 검정 통계치 ===")
print(f"카이제곱 통계량 : {chi2_stat:.4f}")
print(f"자유도 (df)     : {dof}")
print(f"p-value         : {p_val:.4e}")

# 6. 기대빈도 중 5 미만인 칸 수 확인
under_5_count = (expected < 5).sum()
print(f"\n기대빈도 5 미만 칸 수: {under_5_count}개")

# 7. 행 기준 정규화 비율 (구간별 냉방시설 비율)
row_ratio = pd.crosstab(
    df["YearBuiltGroup"], df["CentralAir"], normalize="index"
)
print("\n=== 4. 행 비율 (YearBuiltGroup별 CentralAir 비율) ===")
print((row_ratio * 100).round(2).astype(str) + "%")

# 8. 판정
alpha = 0.05
print("\n=== 5. 판정 ===")
if p_val < alpha:
    print(
        f"p-value({p_val:.4e}) < {alpha} 이므로 귀무가설(H₀)을 기각합니다 (통계적으로 유의한 관련성 있음)."
    )
else:
    print(
        f"p-value({p_val:.4e}) >= {alpha} 이므로 귀무가설(H₀)을 기각하지 못합니다."
    )

=== 1. 교차표 (관측빈도) ===
CentralAir       N    Y
YearBuiltGroup         
1980년 이전        95  744
1980~1999년       0  208
2000년 이후         0  413

=== 2. 기대빈도 표 ===
CentralAir          N       Y
YearBuiltGroup               
1980년 이전        54.59  784.41
1980~1999년      13.53  194.47
2000년 이후        26.87  386.13

=== 3. 카이제곱 검정 통계치 ===
카이제곱 통계량 : 75.2096
자유도 (df)     : 2
p-value         : 4.6605e-17

기대빈도 5 미만 칸 수: 0개

=== 4. 행 비율 (YearBuiltGroup별 CentralAir 비율) ===
CentralAir           N       Y
YearBuiltGroup                
1980년 이전        11.32%  88.68%
1980~1999년        0.0%  100.0%
2000년 이후          0.0%  100.0%

=== 5. 판정 ===
p-value(4.6605e-17) < 0.05 이므로 귀무가설(H₀)을 기각합니다 (통계적으로 유의한 관련성 있음).


### 필수 2 답변 작성란

**Q1.** 이 문제는 적합도 검정과 독립성 검정 중 무엇을 사용해야 하나요?  
- 건축연도와 냉방시설 유무라는 두 범주형 변수의 관계확인, 독립성 검정 사용

**Q2.** 자유도는 얼마이며 어떻게 계산되나요?  
- 3행 2열 교차표, (3-1) * (2-1) = 2 

**Q3.** 기대빈도 조건은 충족되나요?  
- 네, 5미만인 기대 빈도가 0칸으로 기대빈도 기준을 만족.

**Q4.** 건축연도 구간과 중앙 냉방시설 유무 사이에는 유의한 관련성이 있나요?  
- 있습니다. p-value가 0.05보다 작아 두 변수가 독립적이라는 귀무가설을 기각한다.

**Q5.** 카이제곱 검정 결과만으로 건축연도가 냉방시설 설치의 원인이라고 결론 내릴 수 있나요?
- 아니오. 카이제곱 독립성 검정은 관련성을 확인할 뿐, 인과관계를 증명하지 못한다.

---

## 과제. 주방 품질과 진입로 포장 상태의 관계

### 문제 3-1. 두 범주형 변수 재분류 후 독립성 검정

#### 문제 설명

기대빈도가 너무 작은 범주를 줄이기 위해 주방 품질과 진입로 포장 상태를 다음과 같이 재분류합니다.

- `KitchenGroup`
  - 우수: `Ex`, `Gd`
  - 보통 이하: `TA`, `Fa`
- `DriveGroup`
  - 완전 포장: `Y`
  - 미포장·부분포장: `N`, `P`

#### 요구사항

1. 위 기준으로 `KitchenGroup`과 `DriveGroup`을 만드세요.
2. 두 변수의 교차표를 작성하세요.
3. 두 변수가 독립이라는 귀무가설과 관련이 있다는 대립가설을 작성하세요.
4. 카이제곱 독립성 검정을 수행하세요.
5. 카이제곱 통계량, p-value, 자유도와 기대빈도를 출력하세요.
6. 모든 기대빈도가 5 이상인지 확인하세요.
7. 주방 품질 집단별 진입로 포장 비율을 계산하세요.
8. 검정 결과와 비율 차이를 함께 사용하여 두 변수의 관련성을 해석하세요.

#### 해석 질문

**Q1.** 이 과제에서 범주를 재분류한 이유는 무엇인가요?  
**Q2.** 기대빈도 조건은 충족되나요?  
**Q3.** 주방 품질 집단과 진입로 포장 상태는 서로 독립이라고 볼 수 있나요?  
**Q4.** 두 주방 품질 집단의 완전 포장 비율은 각각 얼마인가요?

#### 제출 결과

- 재분류 코드와 교차표
- 가설 설정
- 카이제곱 검정 결과
- 기대빈도 조건 확인
- 행 비율과 결과 해석
- Q1~Q4 답변

In [5]:
# 과제 코드를 작성하세요.
import numpy as np
import pandas as pd
from scipy import stats

# 1. 두 변수 결측치 제외 및 재분류(KitchenGroup, DriveGroup)
df_clean = df[["KitchenQual", "PavedDrive"]].dropna().copy()

kitchen_map = {"Ex": "우수", "Gd": "우수", "TA": "보통 이하", "Fa": "보통 이하"}
drive_map = {"Y": "완전 포장", "N": "미포장·부분포장", "P": "미포장·부분포장"}

df_clean["KitchenGroup"] = df_clean["KitchenQual"].map(kitchen_map)
df_clean["DriveGroup"] = df_clean["PavedDrive"].map(drive_map)

# 2. 교차표 작성 (주변합 포함 및 미포함)
crosstab_raw = pd.crosstab(
    df_clean["KitchenGroup"], df_clean["DriveGroup"]
)
crosstab_margins = pd.crosstab(
    df_clean["KitchenGroup"], df_clean["DriveGroup"], margins=True
)

# 4. 카이제곱 독립성 검정 수행 (correction=False: 수동 통계량과 일관성 유지)
chi2_stat, p_val, dof, expected = stats.chi2_contingency(
    crosstab_raw.values, correction=False
)
expected_df = pd.DataFrame(
    expected, index=crosstab_raw.index, columns=crosstab_raw.columns
)

# 6. 기대빈도 조건 점검 (모든 셀 >= 5)
all_ge_5 = np.all(expected >= 5)

# 7. 주방 품질 집단별 진입로 포장 비율 (행 비율, Row Proportion)
row_props = pd.crosstab(
    df_clean["KitchenGroup"], df_clean["DriveGroup"], normalize="index"
)

# 출력
print("=== 1. 관측 교차표 (주변합 포함) ===")
print(crosstab_margins)

print("\n=== 2. 기대빈도 ===")
print(expected_df.round(2))
print(f"모든 기대빈도 >= 5 충족 여부: {all_ge_5} (최소 기대빈도: {expected.min():.2f})")

print("\n=== 3. 카이제곱 검정 결과 ===")
print(f"카이제곱 통계량 : {chi2_stat:.4f}")
print(f"자유도 (df)     : {dof}")
print(f"p-value         : {p_val:.4e}")

print("\n=== 4. 주방 품질 집단별 포장 상태 비율 (행 비율) ===")
print((row_props * 100).round(2).astype(str) + "%")

=== 1. 관측 교차표 (주변합 포함) ===
DriveGroup    미포장·부분포장  완전 포장   All
KitchenGroup                       
보통 이하              100    674   774
우수                  20    666   686
All                120   1340  1460

=== 2. 기대빈도 ===
DriveGroup    미포장·부분포장   완전 포장
KitchenGroup                  
보통 이하            63.62  710.38
우수               56.38  629.62
모든 기대빈도 >= 5 충족 여부: True (최소 기대빈도: 56.38)

=== 3. 카이제곱 검정 결과 ===
카이제곱 통계량 : 48.2523
자유도 (df)     : 1
p-value         : 3.7476e-12

=== 4. 주방 품질 집단별 포장 상태 비율 (행 비율) ===
DriveGroup   미포장·부분포장   완전 포장
KitchenGroup                 
보통 이하          12.92%  87.08%
우수              2.92%  97.08%


### 과제 답변 작성란

**Q1.** 이 과제에서 범주를 재분류한 이유는 무엇인가요?  
- Fa 주택 수가 너무 적어서, 기대빈도가 너무 낮은 경우가 발생하면 비정상적으로 낮은 p-value가 나올 수 있기 때문.

**Q2.** 기대빈도 조건은 충족되나요?  
- 네 충족합니다. 모든 기대빈도가 5 이상입니다.

**Q3.** 주방 품질 집단과 진입로 포장 상태는 서로 독립이라고 볼 수 있나요?  
- 아니오. p-value가 유의수준 0.05에 비해 매우 작은 수치이므로 두 집단이 독립이라는 귀무가설이 기각됩니다.

**Q4.** 두 주방 품질 집단의 완전 포장 비율은 각각 얼마인가요?
- 보통 이하: 87.08%, 우수: 97.08%

## 실습 마무리

1. 교차표에서 관측빈도와 기대빈도는 어떻게 다른가요?
- 관측빈도는 데이터에서 실제로 센 개수, 기대빈도는 두 변수가 독립이라고 가정할 때 주변합으로 예산되는 개수

2. 기대빈도는 어떤 공식으로 계산하나요?
- (해당 행 x 해당 열) % 전체 합계 

3. 카이제곱 통계량이 커진다는 것은 무엇을 의미하나요?
- 관측빈도가 독립을 가정한 기대빈도에서 전체적으로 더 크게 벗어난다.

4. 독립성 검정과 적합도 검정은 변수 개수와 질문에서 어떻게 다른가요?
- 독립성 검정은 두 범주형 변수의 관계를 확인하고, 
- 적합도 검정은 하나의 범주형 변수의 관측 분포가 기대한 분포와 일치하는지 확인한다.

5. 기대빈도가 5보다 작은 칸이 있다면 무엇을 고려해야 하나요?
- 기대빈도가 작다면 카이제곱 분포로 근사한 p-value의 정확성이 떨어질 수 있다.